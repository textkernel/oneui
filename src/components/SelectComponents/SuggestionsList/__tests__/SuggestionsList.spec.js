import React from 'react';
import toJson from 'enzyme-to-json';
import { SuggestionsList } from '..';
import {
    SUGGESTIONS,
    SUGGESTION_TO_STRING,
} from '../../../AutosuggestDeprecated/__mocks__/suggestions';

const noSuggestionsPlaceholder = 'noSuggestionsPlaceholder';
const getItemProps = (props) => ({ disabled: props.disabled });
const highlightedIndex = 1;

let wrapper;

beforeEach(() => {
    wrapper = mount(
        <SuggestionsList
            isLoading={false}
            suggestions={SUGGESTIONS}
            suggestionToString={SUGGESTION_TO_STRING}
            noSuggestionsPlaceholder={noSuggestionsPlaceholder}
            getItemProps={getItemProps}
            highlightedIndex={highlightedIndex}
        />
    );
});

describe('SuggestionsList', () => {
    it('should initially render component correctly', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should render empty component correctly', () => {
        wrapper.setProps({ suggestions: [] });
        expect(wrapper.find('li')).toHaveLength(1);
    });
    it('should render optimization list correctly', () => {
        wrapper.setProps({ useOptimizeRender: true });
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should render noSuggestions placeholder when empty suggestions list is passed', () => {
        wrapper.setProps({ suggestions: [] });

        expect(wrapper.find('li').childAt(0).text()).toEqual(noSuggestionsPlaceholder);
    });
    it('should render isLoading state', () => {
        wrapper.setProps({ isLoading: true });
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should run suggestionItemRenderer function correctly', () => {
        // eslint-disable-next-line react/display-name
        const suggestionItemRenderer = () => <span>some item text</span>;
        wrapper.setProps({ suggestionItemRenderer });
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should render mixed suggestions and loading list if requested', () => {
        const suggestionsList = SUGGESTIONS.slice(1, 3);
        wrapper.setProps({
            suggestions: suggestionsList,
            isLoading: true,
            allowMixingSuggestionsAndLoading: true,
        });

        expect(wrapper.find('ListItem')).toHaveLength(7);
        expect(wrapper.find('.SuggestionsList__loaderItem')).toHaveLength(5);
    });
    it('should interpret disabled items correctly', () => {
        // assuming last item in suggestions is disabled
        expect(wrapper.find('ListItem').first().prop('disabled')).toBeFalsy();
        expect(wrapper.find('ListItem').last().prop('disabled')).toBeTruthy();
    });
    it('should set li for disabled items as such when prop is set', () => {
        wrapper.setProps({ passDisabledToListItems: true });
        // assuming last item in suggestions is disabled
        expect(wrapper.find('li').first().prop('disabled')).toBeFalsy();
        expect(wrapper.find('li').last().prop('disabled')).toBeTruthy();
    });
});
