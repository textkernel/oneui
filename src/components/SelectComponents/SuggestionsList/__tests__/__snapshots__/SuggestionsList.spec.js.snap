// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuggestionsList should initially render component correctly 1`] = `
<SuggestionsList
  allowMixingSuggestionsAndLoading={false}
  getItemProps={[Function]}
  highlightedIndex={1}
  inputValue=""
  isLoading={false}
  noSuggestionsPlaceholder="noSuggestionsPlaceholder"
  passDisabledToListItems={false}
  suggestionToString={[Function]}
  suggestions={
    Array [
      Object {
        "name": "Sun",
      },
      Object {
        "name": "Moon",
      },
      Object {
        "name": "Mercury",
      },
      Object {
        "name": "Venus",
      },
      Object {
        "name": "Earth",
      },
      Object {
        "name": "Mars",
      },
      Object {
        "name": "Jupiter",
      },
      Object {
        "name": "Saturn",
      },
      Object {
        "name": "Neptune",
      },
      Object {
        "disabled": true,
        "name": "Pluto",
      },
    ]
  }
  useOptimizeRender={false}
>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Sun"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem"
      role="presentation"
      style={Object {}}
    >
      <MarkedText
        inline={true}
        key=".0"
        marker=""
      >
        <Text
          context="default"
          inline={true}
          size="normal"
        >
          <span>
            Sun
          </span>
        </Text>
      </MarkedText>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={true}
    isSelected={false}
    key="Moon"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem ListItem--isHighlighted"
      role="presentation"
      style={Object {}}
    >
      <MarkedText
        inline={true}
        key=".0"
        marker=""
      >
        <Text
          context="default"
          inline={true}
          size="normal"
        >
          <span>
            Moon
          </span>
        </Text>
      </MarkedText>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Mercury"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem"
      role="presentation"
      style={Object {}}
    >
      <MarkedText
        inline={true}
        key=".0"
        marker=""
      >
        <Text
          context="default"
          inline={true}
          size="normal"
        >
          <span>
            Mercury
          </span>
        </Text>
      </MarkedText>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Venus"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem"
      role="presentation"
      style={Object {}}
    >
      <MarkedText
        inline={true}
        key=".0"
        marker=""
      >
        <Text
          context="default"
          inline={true}
          size="normal"
        >
          <span>
            Venus
          </span>
        </Text>
      </MarkedText>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Earth"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem"
      role="presentation"
      style={Object {}}
    >
      <MarkedText
        inline={true}
        key=".0"
        marker=""
      >
        <Text
          context="default"
          inline={true}
          size="normal"
        >
          <span>
            Earth
          </span>
        </Text>
      </MarkedText>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Mars"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem"
      role="presentation"
      style={Object {}}
    >
      <MarkedText
        inline={true}
        key=".0"
        marker=""
      >
        <Text
          context="default"
          inline={true}
          size="normal"
        >
          <span>
            Mars
          </span>
        </Text>
      </MarkedText>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Jupiter"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem"
      role="presentation"
      style={Object {}}
    >
      <MarkedText
        inline={true}
        key=".0"
        marker=""
      >
        <Text
          context="default"
          inline={true}
          size="normal"
        >
          <span>
            Jupiter
          </span>
        </Text>
      </MarkedText>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Saturn"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem"
      role="presentation"
      style={Object {}}
    >
      <MarkedText
        inline={true}
        key=".0"
        marker=""
      >
        <Text
          context="default"
          inline={true}
          size="normal"
        >
          <span>
            Saturn
          </span>
        </Text>
      </MarkedText>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Neptune"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem"
      role="presentation"
      style={Object {}}
    >
      <MarkedText
        inline={true}
        key=".0"
        marker=""
      >
        <Text
          context="default"
          inline={true}
          size="normal"
        >
          <span>
            Neptune
          </span>
        </Text>
      </MarkedText>
    </li>
  </ListItem>
  <ListItem
    disabled={true}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Pluto"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem ListItem--disabled"
      role="presentation"
      style={Object {}}
    >
      <MarkedText
        inline={true}
        key=".0"
        marker=""
      >
        <Text
          context="default"
          inline={true}
          size="normal"
        >
          <span>
            Pluto
          </span>
        </Text>
      </MarkedText>
    </li>
  </ListItem>
</SuggestionsList>
`;

exports[`SuggestionsList should render isLoading state 1`] = `
<SuggestionsList
  allowMixingSuggestionsAndLoading={false}
  getItemProps={[Function]}
  highlightedIndex={1}
  inputValue=""
  isLoading={true}
  noSuggestionsPlaceholder="noSuggestionsPlaceholder"
  passDisabledToListItems={false}
  suggestionToString={[Function]}
  suggestions={
    Array [
      Object {
        "name": "Sun",
      },
      Object {
        "name": "Moon",
      },
      Object {
        "name": "Mercury",
      },
      Object {
        "name": "Venus",
      },
      Object {
        "name": "Earth",
      },
      Object {
        "name": "Mars",
      },
      Object {
        "name": "Jupiter",
      },
      Object {
        "name": "Saturn",
      },
      Object {
        "name": "Neptune",
      },
      Object {
        "disabled": true,
        "name": "Pluto",
      },
    ]
  }
  useOptimizeRender={false}
>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="0"
    passDisabledToLi={false}
  >
    <li
      className="ListItem"
      role="presentation"
    >
      <div
        className="SuggestionsList__loaderItem"
        key=".0"
      >
        <ContentPlaceholder
          duration={1}
          width={100}
          withoutMargin={false}
        >
          <div
            className="ContentPlaceholder"
            style={
              Object {
                "animationDuration": "1s",
                "height": undefined,
              }
            }
          >
             
            <div
              className="ContentPlaceholder__mask"
              style={
                Object {
                  "width": "0%",
                }
              }
            />
          </div>
        </ContentPlaceholder>
      </div>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="1"
    passDisabledToLi={false}
  >
    <li
      className="ListItem"
      role="presentation"
    >
      <div
        className="SuggestionsList__loaderItem"
        key=".0"
      >
        <ContentPlaceholder
          duration={1}
          width={100}
          withoutMargin={false}
        >
          <div
            className="ContentPlaceholder"
            style={
              Object {
                "animationDuration": "1s",
                "height": undefined,
              }
            }
          >
             
            <div
              className="ContentPlaceholder__mask"
              style={
                Object {
                  "width": "0%",
                }
              }
            />
          </div>
        </ContentPlaceholder>
      </div>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="2"
    passDisabledToLi={false}
  >
    <li
      className="ListItem"
      role="presentation"
    >
      <div
        className="SuggestionsList__loaderItem"
        key=".0"
      >
        <ContentPlaceholder
          duration={1}
          width={100}
          withoutMargin={false}
        >
          <div
            className="ContentPlaceholder"
            style={
              Object {
                "animationDuration": "1s",
                "height": undefined,
              }
            }
          >
             
            <div
              className="ContentPlaceholder__mask"
              style={
                Object {
                  "width": "0%",
                }
              }
            />
          </div>
        </ContentPlaceholder>
      </div>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="3"
    passDisabledToLi={false}
  >
    <li
      className="ListItem"
      role="presentation"
    >
      <div
        className="SuggestionsList__loaderItem"
        key=".0"
      >
        <ContentPlaceholder
          duration={1}
          width={100}
          withoutMargin={false}
        >
          <div
            className="ContentPlaceholder"
            style={
              Object {
                "animationDuration": "1s",
                "height": undefined,
              }
            }
          >
             
            <div
              className="ContentPlaceholder__mask"
              style={
                Object {
                  "width": "0%",
                }
              }
            />
          </div>
        </ContentPlaceholder>
      </div>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="4"
    passDisabledToLi={false}
  >
    <li
      className="ListItem"
      role="presentation"
    >
      <div
        className="SuggestionsList__loaderItem"
        key=".0"
      >
        <ContentPlaceholder
          duration={1}
          width={100}
          withoutMargin={false}
        >
          <div
            className="ContentPlaceholder"
            style={
              Object {
                "animationDuration": "1s",
                "height": undefined,
              }
            }
          >
             
            <div
              className="ContentPlaceholder__mask"
              style={
                Object {
                  "width": "0%",
                }
              }
            />
          </div>
        </ContentPlaceholder>
      </div>
    </li>
  </ListItem>
</SuggestionsList>
`;

exports[`SuggestionsList should render optimization list correctly 1`] = `
<SuggestionsList
  allowMixingSuggestionsAndLoading={false}
  getItemProps={[Function]}
  highlightedIndex={1}
  inputValue=""
  isLoading={false}
  noSuggestionsPlaceholder="noSuggestionsPlaceholder"
  passDisabledToListItems={false}
  suggestionToString={[Function]}
  suggestions={
    Array [
      Object {
        "name": "Sun",
      },
      Object {
        "name": "Moon",
      },
      Object {
        "name": "Mercury",
      },
      Object {
        "name": "Venus",
      },
      Object {
        "name": "Earth",
      },
      Object {
        "name": "Mars",
      },
      Object {
        "name": "Jupiter",
      },
      Object {
        "name": "Saturn",
      },
      Object {
        "name": "Neptune",
      },
      Object {
        "disabled": true,
        "name": "Pluto",
      },
    ]
  }
  useOptimizeRender={true}
>
  <div
    className="SuggestionsList__optimizerWrapper"
  >
    <ListOptimizer
      rowCount={10}
    >
      <AutoSizer
        disableHeight={false}
        disableWidth={false}
        onResize={[Function]}
        style={Object {}}
      >
        <div
          style={
            Object {
              "height": 0,
              "overflow": "visible",
              "width": 0,
            }
          }
        >
          <List
            autoHeight={false}
            deferredMeasurementCache={
              CellMeasurerCache {
                "_cellHeightCache": Object {},
                "_cellWidthCache": Object {},
                "_columnCount": 0,
                "_columnWidthCache": Object {},
                "_defaultHeight": 30,
                "_defaultWidth": 100,
                "_hasFixedHeight": false,
                "_hasFixedWidth": true,
                "_keyMapper": [Function],
                "_minHeight": 0,
                "_minWidth": 0,
                "_rowCount": 0,
                "_rowHeightCache": Object {},
                "columnWidth": [Function],
                "rowHeight": [Function],
              }
            }
            estimatedRowSize={30}
            height={0}
            noRowsRenderer={[Function]}
            onRowsRendered={[Function]}
            onScroll={[Function]}
            overscanIndicesGetter={[Function]}
            overscanRowCount={10}
            rowCount={10}
            rowHeight={[Function]}
            rowRenderer={[Function]}
            scrollToAlignment="auto"
            scrollToIndex={-1}
            style={Object {}}
            width={0}
          >
            <Grid
              aria-label="grid"
              aria-readonly={true}
              autoContainerWidth={true}
              autoHeight={false}
              autoWidth={false}
              cellRangeRenderer={[Function]}
              cellRenderer={[Function]}
              className="ReactVirtualized__List"
              columnCount={1}
              columnWidth={0}
              containerRole="rowgroup"
              containerStyle={Object {}}
              deferredMeasurementCache={
                CellMeasurerCache {
                  "_cellHeightCache": Object {},
                  "_cellWidthCache": Object {},
                  "_columnCount": 0,
                  "_columnWidthCache": Object {},
                  "_defaultHeight": 30,
                  "_defaultWidth": 100,
                  "_hasFixedHeight": false,
                  "_hasFixedWidth": true,
                  "_keyMapper": [Function],
                  "_minHeight": 0,
                  "_minWidth": 0,
                  "_rowCount": 0,
                  "_rowHeightCache": Object {},
                  "columnWidth": [Function],
                  "rowHeight": [Function],
                }
              }
              estimatedColumnSize={100}
              estimatedRowSize={30}
              getScrollbarSize={[Function]}
              height={0}
              isScrollingOptOut={false}
              noContentRenderer={[Function]}
              noRowsRenderer={[Function]}
              onRowsRendered={[Function]}
              onScroll={[Function]}
              onScrollbarPresenceChange={[Function]}
              onSectionRendered={[Function]}
              overscanColumnCount={0}
              overscanIndicesGetter={[Function]}
              overscanRowCount={10}
              role="grid"
              rowCount={10}
              rowHeight={[Function]}
              rowRenderer={[Function]}
              scrollToAlignment="auto"
              scrollToColumn={-1}
              scrollToIndex={-1}
              scrollToRow={-1}
              scrollingResetTimeInterval={150}
              style={Object {}}
              tabIndex={0}
              width={0}
            >
              <div
                aria-label="grid"
                aria-readonly={true}
                className="ReactVirtualized__Grid ReactVirtualized__List"
                onScroll={[Function]}
                role="grid"
                style={
                  Object {
                    "WebkitOverflowScrolling": "touch",
                    "boxSizing": "border-box",
                    "direction": "ltr",
                    "height": 0,
                    "overflowX": "hidden",
                    "overflowY": "auto",
                    "position": "relative",
                    "width": 0,
                    "willChange": "transform",
                  }
                }
                tabIndex={0}
              />
            </Grid>
          </List>
        </div>
      </AutoSizer>
    </ListOptimizer>
  </div>
</SuggestionsList>
`;

exports[`SuggestionsList should run suggestionItemRenderer function correctly 1`] = `
<SuggestionsList
  allowMixingSuggestionsAndLoading={false}
  getItemProps={[Function]}
  highlightedIndex={1}
  inputValue=""
  isLoading={false}
  noSuggestionsPlaceholder="noSuggestionsPlaceholder"
  passDisabledToListItems={false}
  suggestionItemRenderer={[Function]}
  suggestionToString={[Function]}
  suggestions={
    Array [
      Object {
        "name": "Sun",
      },
      Object {
        "name": "Moon",
      },
      Object {
        "name": "Mercury",
      },
      Object {
        "name": "Venus",
      },
      Object {
        "name": "Earth",
      },
      Object {
        "name": "Mars",
      },
      Object {
        "name": "Jupiter",
      },
      Object {
        "name": "Saturn",
      },
      Object {
        "name": "Neptune",
      },
      Object {
        "disabled": true,
        "name": "Pluto",
      },
    ]
  }
  useOptimizeRender={false}
>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Sun"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem"
      role="presentation"
      style={Object {}}
    >
      <span
        key=".0"
      >
        some item text
      </span>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={true}
    isSelected={false}
    key="Moon"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem ListItem--isHighlighted"
      role="presentation"
      style={Object {}}
    >
      <span
        key=".0"
      >
        some item text
      </span>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Mercury"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem"
      role="presentation"
      style={Object {}}
    >
      <span
        key=".0"
      >
        some item text
      </span>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Venus"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem"
      role="presentation"
      style={Object {}}
    >
      <span
        key=".0"
      >
        some item text
      </span>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Earth"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem"
      role="presentation"
      style={Object {}}
    >
      <span
        key=".0"
      >
        some item text
      </span>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Mars"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem"
      role="presentation"
      style={Object {}}
    >
      <span
        key=".0"
      >
        some item text
      </span>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Jupiter"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem"
      role="presentation"
      style={Object {}}
    >
      <span
        key=".0"
      >
        some item text
      </span>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Saturn"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem"
      role="presentation"
      style={Object {}}
    >
      <span
        key=".0"
      >
        some item text
      </span>
    </li>
  </ListItem>
  <ListItem
    disabled={false}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Neptune"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem"
      role="presentation"
      style={Object {}}
    >
      <span
        key=".0"
      >
        some item text
      </span>
    </li>
  </ListItem>
  <ListItem
    disabled={true}
    highlightContext="default"
    isHighlighted={false}
    isSelected={false}
    key="Pluto"
    passDisabledToLi={false}
    style={Object {}}
  >
    <li
      className="ListItem ListItem--disabled"
      role="presentation"
      style={Object {}}
    >
      <span
        key=".0"
      >
        some item text
      </span>
    </li>
  </ListItem>
</SuggestionsList>
`;
